In [1]:
import numpy as np 
import pandas as pd 

%load_ext autoreload
%autoreload 2
import funciones_etl
import conexion_bd
import warnings
warnings.filterwarnings("ignore")

### Se extraen los datos, pestaña HECHOS

In [2]:
# Abrimos el archivo excel
homicidios_hechos = pd.read_excel("../Datos/homicidios.xlsx", sheet_name="HECHOS")
homicidios_hechos

ID  N_VICTIMAS      FECHA  AAAA  MM  DD      HORA  HH  \
0    2016-0001           1 2016-01-01  2016   1   1  04:00:00   4   
1    2016-0002           1 2016-01-02  2016   1   2  01:15:00   1   
2    2016-0003           1 2016-01-03  2016   1   3  07:00:00   7   
3    2016-0004           1 2016-01-10  2016   1  10  00:00:00   0   
4    2016-0005           1 2016-01-21  2016   1  21  05:20:00   5   
..         ...         ...        ...   ...  ..  ..       ...  ..   
691  2021-0093           1 2021-12-13  2021  12  13  17:10:00  17   
692  2021-0094           1 2021-12-20  2021  12  20  01:10:00   1   
693  2021-0095           1 2021-12-30  2021  12  30  00:43:00   0   
694  2021-0096           1 2021-12-15  2021  12  15  10:30:00  10   
695  2021-0097           1 2021-11-18  2021  11  18  06:10:00   6   

                               LUGAR_DEL_HECHO TIPO_DE_CALLE  ...  Altura  \
0    AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ       AVENIDA  ...     NaN   
1            AV GRAL PAZ  Y AV DE LOS CORRALES      GRAL PAZ  ...     NaN   
2                           AV ENTRE RIOS 2034       AVENIDA  ...  2034.0   
3        AV LARRAZABAL Y GRAL VILLEGAS CONRADO       AVENIDA  ...     NaN   
4    AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA        AVENIDA  ...     NaN   
..                                         ...           ...  ...     ...   
691                          AV. RIESTRA Y MOM       AVENIDA  ...     NaN   
692                    AU DELLEPIANE Y LACARRA     AUTOPISTA  ...     NaN   
693                        AV. GAONA Y TERRADA       AVENIDA  ...     NaN   
694                         AV. EVA PERON 4071       AVENIDA  ...  4071.0   
695                    PADRE CARLOS MUJICA 709         CALLE  ...   709.0   

                                   Cruce  \
0    FERNANDEZ DE LA CRUZ, F., GRAL. AV.   
1                    DE LOS CORRALES AV.   
2                                    NaN   
3               VILLEGAS, CONRADO, GRAL.   
4                SAENZ PE?A, LUIS, PRES.   
..                                   ...   
691                                  MOM   
692                          LACARRA AV.   
693                              TERRADA   
694                                  NaN   
695                                  NaN   

                                 Dirección Normalizada COMUNA  \
0    PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...      8   
1                 PAZ, GRAL. AV. y DE LOS CORRALES AV.      9   
2                                  ENTRE RIOS AV. 2034      1   
3            LARRAZABAL AV. y VILLEGAS, CONRADO, GRAL.      8   
4              SAN JUAN AV. y SAENZ PEÑA, LUIS, PRES.      1   
..                                                 ...    ...   
691                                  RIESTRA AV. y MOM      7   
692         DELLEPIANE, LUIS, TTE. GRAL. y LACARRA AV.      9   
693                                GAONA AV. y TERRADA     11   
694                                PERON, EVA AV. 4071      9   
695                            PADRE CARLOS MUGICA 709      1   

                                   XY (CABA)         pos x         pos y  \
0      Point (98896.78238426 93532.43437792)  -58.47533969  -34.68757022   
1      Point (95832.05571093 95505.41641999)  -58.50877521  -34.66977709   
2     Point (106684.29090040 99706.57687843)  -58.39040293  -34.63189362   
3      Point (99840.65224780 94269.16534422)  -58.46503904  -34.68092974   
4    Point (106980.32827929 100752.16915795)  -58.38718297  -34.62246630   
..                                       ...           ...           ...   
691   Point (102728.60090138 98186.24929177)  -58.43353773  -34.64561636   
692    Point (99624.29795829 97569.69801131)  -58.46739825  -34.65117757   
693   Point (99116.45492358 101045.23284826)  -58.47293407  -34.61984745   
694    Point (99324.54463985 97676.26932409)  -58.47066794  -34.65021673   
695  Point (107664.16647795 104708.63962087)  -58.37976155  -34.58679619   

      PARTICIPANTES    VICTIMA    A

Entre los datos generales, una vez abierto este archivo podemos encontrar que tiene 21 columnas y 696 filas

In [3]:
# Observar columnas
homicidios_hechos.columns

Index(['ID', 'N_VICTIMAS', 'FECHA', 'AAAA', 'MM', 'DD', 'HORA', 'HH',
       'LUGAR_DEL_HECHO', 'TIPO_DE_CALLE', 'Calle', 'Altura', 'Cruce',
       'Dirección Normalizada', 'COMUNA', 'XY (CABA)', 'pos x', 'pos y',
       'PARTICIPANTES', 'VICTIMA', 'ACUSADO'],
      dtype='object')

In [4]:
# Observar cantidad de registros o filas
homicidios_hechos.shape[0]

696

Se puede corroborar la cantidad de columnas y de registros en este dataframe

### Normalizar encabezados de las columnas

Como algunas columnas tienen su nombre en mayusaculas otras en minusculas ademas de que unas cuentan con espacios y otras con "_" en su lugas decidimos estandarizar dichas columnas, de manera que eliminiaremos los esapacios pensando en cargar la data en una base de datos con motor mysql

In [5]:
# Capitalize, para colocar la primera letra en mayuscula
homicidios_hechos.columns = [x.capitalize() for x in homicidios_hechos.columns]

#Se reemplazan los espacios por guion bajo "-"
homicidios_hechos.columns = homicidios_hechos.columns.str.replace(" ", "_")

# Se renombran algunas columnas para darle mas sentido al dataframe
homicidios_hechos = homicidios_hechos.rename(columns={"Id": "Id_hecho",
                                                      "N_victimas": "Cantidad_victimas",
                                                      "Aaaa": "Año",
                                                      "Mm": "Mes",
                                                      "Dd": "Dia",
                                                      "Hh": "Hora_entero"})

In [6]:
homicidios_hechos.columns

Index(['Id_hecho', 'Cantidad_victimas', 'Fecha', 'Año', 'Mes', 'Dia', 'Hora',
       'Hora_entero', 'Lugar_del_hecho', 'Tipo_de_calle', 'Calle', 'Altura',
       'Cruce', 'Dirección_normalizada', 'Comuna', 'Xy_(caba)', 'Pos_x',
       'Pos_y', 'Participantes', 'Victima', 'Acusado'],
      dtype='object')

### Valores nulos

In [7]:
nulos_tipo = funciones_etl.verificar_datos(homicidios_hechos)
nulos_tipo

Advertencia: la columna Calle, tiene valores nulos
Advertencia: la columna Altura, tiene valores nulos
Advertencia: la columna Cruce, tiene valores nulos
Advertencia: la columna Dirección_normalizada, tiene valores nulos


columna                                          tipo_dato  \
0                Id_hecho                                    [<class 'str'>]   
1       Cantidad_victimas                                    [<class 'int'>]   
2                   Fecha  [<class 'pandas._libs.tslibs.timestamps.Timest...   
3                     Año                                    [<class 'int'>]   
4                     Mes                                    [<class 'int'>]   
5                     Dia                                    [<class 'int'>]   
6                    Hora  [<class 'datetime.time'>, <class 'str'>, <clas...   
7             Hora_entero                     [<class 'int'>, <class 'str'>]   
8         Lugar_del_hecho                                    [<class 'str'>]   
9           Tipo_de_calle                                    [<class 'str'>]   
10                  Calle                   [<class 'str'>, <class 'float'>]   
11                 Altura                                  [<class 'float'>]   
12                  Cruce                   [<class 'str'>, <class 'float'>]   
13  Dirección_normalizada                   [<class 'str'>, <class 'float'>]   
14                 Comuna                                    [<class 'int'>]   
15              Xy_(caba)                                    [<class 'str'>]   
16                  Pos_x                                    [<class 'str'>]   
17                  Pos_y                                    [<class 'str'>]   
18          Participantes                                    [<class 'str'>]   
19                Victima                                    [<class 'str'>]   
20                Acusado                                    [<class 'str'>]   

    datos_nulos  porcentaje_nulos  porcentaje_no_nulos  
0             0              0.00               100.00  
1             0              0.00               100.00  
2             0              0.00               100.00  
3             0              0.00               100.00  
4             0              0.00               100.00  
5             0              0.00               100.00  
6             0              0.00               100.00  
7             0              0.00               100.00  
8             0              0.00               100.00  
9             0              0.00               100.00  
10            1              0.14                99.86  
11          567             81.47                18.53  
12          171             24.57                75.43  
13            8              1.15                98.85  
14            0              0.00               100.00  
15            0              0.00               100.00  
16            0              0.00               100.00  
17            0              0.00               100.00  
18            0              0.00               100.00  
19            0              0.00               100.00  
20            0              0.00               100.00

Podemos observar 747 datos nulos en el dataframe, repartidos de la siguiente manera

* Columna Altura = 567 datos nulos, que representan un 81.47 % de sus datos

* Columna Cruce = 171 datos nulos, que representan 24.57 % de sus datos

* Columna Calle = 1 datos nulos, que representan el 0.14 % de sus datos

* Columna Direccion_normalizada = 8 datos nulos, que representan 1,15 % de sus datos

Altura: 

Comparamos los vaores de las columnas Altura, Lugar_del_hecho y Direcccion_normalizada, esto debido a que las res columnas se trata de datos sobre la direccion donde ocurre el siniestro asi que de esta manera buscamos coincidencias que nos permitan completar informacion

In [8]:
# Comparamos los valores de las columnas
df_no_nulos_altura = homicidios_hechos[["Altura", "Lugar_del_hecho", "Dirección_normalizada"]]
df_no_nulos_altura

Altura                            Lugar_del_hecho  \
0       NaN  AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ   
1       NaN          AV GRAL PAZ  Y AV DE LOS CORRALES   
2    2034.0                         AV ENTRE RIOS 2034   
3       NaN      AV LARRAZABAL Y GRAL VILLEGAS CONRADO   
4       NaN  AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA    
..      ...                                        ...   
691     NaN                          AV. RIESTRA Y MOM   
692     NaN                    AU DELLEPIANE Y LACARRA   
693     NaN                        AV. GAONA Y TERRADA   
694  4071.0                         AV. EVA PERON 4071   
695   709.0                    PADRE CARLOS MUJICA 709   

                                 Dirección_normalizada  
0    PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...  
1                 PAZ, GRAL. AV. y DE LOS CORRALES AV.  
2                                  ENTRE RIOS AV. 2034  
3            LARRAZABAL AV. y VILLEGAS, CONRADO, GRAL.  
4              SAN JUAN AV. y SAENZ PEÑA, LUIS, PRES.  
..                                                 ...  
691                                  RIESTRA AV. y MOM  
692         DELLEPIANE, LUIS, TTE. GRAL. y LACARRA AV.  
693                                GAONA AV. y TERRADA  
694                                PERON, EVA AV. 4071  
695                            PADRE CARLOS MUGICA 709  

[696 rows x 3 columns]

Obtenemos los datos no nulos y los mostramos

In [9]:
no_nulos_altura = df_no_nulos_altura.loc[~df_no_nulos_altura.isnull().any(axis=1)]

In [10]:
no_nulos_altura

Altura          Lugar_del_hecho    Dirección_normalizada
2    2034.0       AV ENTRE RIOS 2034      ENTRE RIOS AV. 2034
9    1366.0       AV ENTRE RIOS 1366      ENTRE RIOS AV. 1366
14    156.0             SUIPACHA 156             SUIPACHA 156
33   1483.0                LIMA 1483                LIMA 1483
42     30.0  AV LUIS MARIA CAMPOS 30   CAMPOS, LUIS M. AV. 30
..      ...                      ...                      ...
677  3112.0             TUCUMAN 3112             TUCUMAN 3112
685  1359.0           ESMERALDA 1359           ESMERALDA 1359
687  3655.0           AV. GAONA 3655           GAONA AV. 3655
694  4071.0       AV. EVA PERON 4071      PERON, EVA AV. 4071
695   709.0  PADRE CARLOS MUJICA 709  PADRE CARLOS MUGICA 709

[129 rows x 3 columns]

Podemos observar que los datos que no son nulos de la columnas Altura se repiten en las columnas Lugar_delhecho y Direccion_normalizada, por lo tanto se decide eliminar la columna

In [11]:
homicidios_hechos = homicidios_hechos.drop("Altura", axis=1)
homicidios_hechos.columns

Index(['Id_hecho', 'Cantidad_victimas', 'Fecha', 'Año', 'Mes', 'Dia', 'Hora',
       'Hora_entero', 'Lugar_del_hecho', 'Tipo_de_calle', 'Calle', 'Cruce',
       'Dirección_normalizada', 'Comuna', 'Xy_(caba)', 'Pos_x', 'Pos_y',
       'Participantes', 'Victima', 'Acusado'],
      dtype='object')

Cruce:

Esta es una variable categorica,de tipo string. con ella podemos evaluar los suceso que van ocurreiendo en las esquonas por ejemplo.

In [12]:
# Observamos los datos nulos de la columna cruce
nulos_cruce = homicidios_hechos[homicidios_hechos["Cruce"].isnull()][["Cruce", "Lugar_del_hecho"]]
nulos_cruce

Cruce                                    Lugar_del_hecho
2     NaN                                 AV ENTRE RIOS 2034
9     NaN                                 AV ENTRE RIOS 1366
14    NaN                                       SUIPACHA 156
33    NaN                                          LIMA 1483
35    NaN  AUTOPISTA 1 SUR PRESIDENTE ARTURO FRONDIZI KM....
..    ...                                                ...
680   NaN                   AU 25 DE MAYO KM7 (ALTURA PASCO)
685   NaN                                     ESMERALDA 1359
687   NaN                                     AV. GAONA 3655
694   NaN                                 AV. EVA PERON 4071
695   NaN                            PADRE CARLOS MUJICA 709

[171 rows x 2 columns]

In [13]:
# Observamos los datos nonulos de la columna cruce
no_nulos_cruce = homicidios_hechos[homicidios_hechos["Cruce"].notnull()][["Cruce", "Lugar_del_hecho"]]
no_nulos_cruce

Cruce  \
0    FERNANDEZ DE LA CRUZ, F., GRAL. AV.   
1                    DE LOS CORRALES AV.   
3               VILLEGAS, CONRADO, GRAL.   
4                SAENZ PE?A, LUIS, PRES.   
5                           ESCALADA AV.   
..                                   ...   
689                         HUGO, VICTOR   
690                       PUEYRREDON AV.   
691                                  MOM   
692                          LACARRA AV.   
693                              TERRADA   

                               Lugar_del_hecho  
0    AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ  
1            AV GRAL PAZ  Y AV DE LOS CORRALES  
3        AV LARRAZABAL Y GRAL VILLEGAS CONRADO  
4    AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA   
5               AV 27 DE FEBRERO Y AV ESCALADA  
..                                         ...  
689                    BAIGORRIA Y VICTOR HUGO  
690             AV. RIVADAVIA Y AV. PUEYRREDON  
691                          AV. RIESTRA Y MOM  
692                    AU DELLEPIANE Y LACARRA  
693                        AV. GAONA Y TERRADA  

[525 rows x 2 columns]

Podemos observar que los datos nulos de la columna cruce se deben a que el eccidente no esta ocurriendo en una esquina o en un cruce, ya que podemos ver la diferencia que hay entre los dos dataframes que nos indica que el de los datos nulos la columna Lugar_del_hecho solo contiene una avenida como direccion y su altura, en cambio en el datafframe de los datos no nulos la esta misma columna contiene doas avenidas o dos calles lo que nos indica en su formato que apunta a un cruce o esquina

In [14]:
# Se reemplazan los valores no nulos por "si" y los nulos por "no"
homicidios_hechos['Cruce'] = np.where(homicidios_hechos['Cruce'].notnull(), 'SI', 'NO')

Se reemplazan los valores para los casos que ocurren en una esquimna por "si" y los que no ocurren en una esquina por un "no", y podemos observar como nos queda la columna

In [15]:
# Mostramos columna cruce
homicidios_hechos[["Cruce", "Lugar_del_hecho"]]

Cruce                            Lugar_del_hecho
0      SI  AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ
1      SI          AV GRAL PAZ  Y AV DE LOS CORRALES
2      NO                         AV ENTRE RIOS 2034
3      SI      AV LARRAZABAL Y GRAL VILLEGAS CONRADO
4      SI  AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA 
..    ...                                        ...
691    SI                          AV. RIESTRA Y MOM
692    SI                    AU DELLEPIANE Y LACARRA
693    SI                        AV. GAONA Y TERRADA
694    NO                         AV. EVA PERON 4071
695    NO                    PADRE CARLOS MUJICA 709

[696 rows x 2 columns]

Direecion Normalizada

Observamos los datos nulos de esta columna

In [16]:
nulos_direccion_norma = homicidios_hechos[homicidios_hechos["Dirección_normalizada"].isnull()][["Dirección_normalizada", "Lugar_del_hecho", "Pos_x", "Pos_y"]]
nulos_direccion_norma

Dirección_normalizada                          Lugar_del_hecho Pos_x Pos_y
38                    NaN               AUTOPISTA LUGONES PK 10000     .     .
106                   NaN         AU BUENOS AIRES - LA PLATA KM. 4     .     .
119                   NaN                                       SD     .     .
180                   NaN  AU PERITO MORENO Y RAMAL ENLACE AU1/AU6     .     .
181                   NaN                       AU DELLEPIANE 2400     .     .
313                   NaN                 AUTOPISTA LUGONES KM 4.7     .     .
546                   NaN             LUGONES, LEOPOLDO AV. KM 6,1     .     .
621                   NaN         AU BUENOS AIRES LA PLATA  KM 4,5     .     .

Podemos observar que no se cuenta con informacion suficiente para amputar estos valores

In [17]:
# Se rellenan nulos con SD
homicidios_hechos['Dirección_normalizada'].fillna("SD", inplace=True)

Como no contamos con informacion para amputar los valotres, se reemplazaron con el valor "SD" "Sin Dato", tomando en cuenta tambien que solo representan el 1.15%

In [18]:
# Mostramos el dataframe
homicidios_hechos[["Cruce", "Dirección_normalizada", "Lugar_del_hecho"]]

Cruce                              Dirección_normalizada  \
0      SI  PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...   
1      SI               PAZ, GRAL. AV. y DE LOS CORRALES AV.   
2      NO                                ENTRE RIOS AV. 2034   
3      SI          LARRAZABAL AV. y VILLEGAS, CONRADO, GRAL.   
4      SI            SAN JUAN AV. y SAENZ PEÑA, LUIS, PRES.   
..    ...                                                ...   
691    SI                                  RIESTRA AV. y MOM   
692    SI         DELLEPIANE, LUIS, TTE. GRAL. y LACARRA AV.   
693    SI                                GAONA AV. y TERRADA   
694    NO                                PERON, EVA AV. 4071   
695    NO                            PADRE CARLOS MUGICA 709   

                               Lugar_del_hecho  
0    AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ  
1            AV GRAL PAZ  Y AV DE LOS CORRALES  
2                           AV ENTRE RIOS 2034  
3        AV LARRAZABAL Y GRAL VILLEGAS CONRADO  
4    AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA   
..                                         ...  
691                          AV. RIESTRA Y MOM  
692                    AU DELLEPIANE Y LACARRA  
693                        AV. GAONA Y TERRADA  
694                         AV. EVA PERON 4071  
695                    PADRE CARLOS MUJICA 709  

[696 rows x 3 columns]

### Valores Duplicados

In [19]:
valores = len(homicidios_hechos["Id_hecho"])
valores

696

In [20]:
funciones_etl.valores_duplicados(homicidios_hechos, "Id_hecho")

0

La columna Id_hecho cuenta con 696 valores en total y tiene 0 valores duplicados

### Tipo de Dato

In [21]:
funciones_etl.verificar_tipo_variable(homicidios_hechos)

nombre_campo                                         tipo_datos
0                Id_hecho                                    [<class 'str'>]
1       Cantidad_victimas                                    [<class 'int'>]
2                   Fecha  [<class 'pandas._libs.tslibs.timestamps.Timest...
3                     Año                                    [<class 'int'>]
4                     Mes                                    [<class 'int'>]
5                     Dia                                    [<class 'int'>]
6                    Hora  [<class 'datetime.time'>, <class 'str'>, <clas...
7             Hora_entero                     [<class 'int'>, <class 'str'>]
8         Lugar_del_hecho                                    [<class 'str'>]
9           Tipo_de_calle                                    [<class 'str'>]
10                  Calle                   [<class 'str'>, <class 'float'>]
11                  Cruce                                    [<class 'str'>]
12  Dirección_normalizada                                    [<class 'str'>]
13                 Comuna                                    [<class 'int'>]
14              Xy_(caba)                                    [<class 'str'>]
15                  Pos_x                                    [<class 'str'>]
16                  Pos_y                                    [<class 'str'>]
17          Participantes                                    [<class 'str'>]
18                Victima                                    [<class 'str'>]
19                Acusado                                    [<class 'str'>]

Las columna Fecha, Hora, Hora_entero y Calle tienen mas de un tipo de datos 

Hora

In [22]:
# Observamos el tipo de dato de la columna hora
homicidios_hechos["Hora"].apply(type).value_counts()

Hora
<class 'datetime.time'>        608
<class 'str'>                   85
<class 'datetime.datetime'>      3
Name: count, dtype: int64

Observamos 608 datos de tipo datatime.time, 85 de tipo string y 3 de tipo datetime.datetime. Se decide dejar todos los datos de tipo datetime.time ya que somn los que mayor cantidad hay 

In [23]:
# convertimos a time
homicidios_hechos["Hora"] = homicidios_hechos["Hora"].apply(lambda x: funciones_etl.convertir_time(x))

In [24]:
# Verificamos los tipos de datos
homicidios_hechos["Hora"].apply(type).value_counts()

Hora
<class 'datetime.time'>    695
<class 'NoneType'>           1
Name: count, dtype: int64

Cambiamos los datos a tipo datetime.time, pero nos sigue faltando uno que no se pudo convertir 

In [25]:
homicidios_hechos[homicidios_hechos["Hora"].isna()]

Id_hecho  Cantidad_victimas      Fecha   Año  Mes  Dia  Hora  \
518  2019-0103                  1 2019-12-18  2019   12   18  None   

    Hora_entero          Lugar_del_hecho Tipo_de_calle           Calle Cruce  \
518          SD  PAZ, GRAL. AV. Y GRIVEO      GRAL PAZ  PAZ, GRAL. AV.    SI   

       Dirección_normalizada  Comuna                               Xy_(caba)  \
518  PAZ, GRAL. AV. y GRIVEO      11  Point (94643.11254058 103831.57115061)   

            Pos_x         Pos_y Participantes Victima Acusado  
518  -58.52169422  -34.59471640     MOTO-MOTO    MOTO    MOTO

Obsrvamos que el hay un dato en le columna Hora que aparece como None, y en la columnas Hora_entero como SD, por lo tanto este puede ser el dato que nos esta faltanto en ambas columnas. Para la columna Hora se decide imputar mediante medidas de tendencia central, especificamente por la moda

In [26]:
# Calculamos la moda
hora_moda = homicidios_hechos["Hora"].mode().iloc[0]

In [27]:
# Reemplazamos
homicidios_hechos["Hora"].fillna(hora_moda, inplace=True)

In [28]:
# Verificamos
homicidios_hechos["Hora"].apply(type).value_counts()

Hora
<class 'datetime.time'>    696
Name: count, dtype: int64

En este proceso calculamos la hora mas comun o la moda de la columna para imputar el valor que aun teniamos como nulo

Hora entero

Se pudo observer cual es el dato que se presenta como nulo en esta columna esta como SD, los cuales vamos a reemplazar por la moda

In [29]:
# Verificamos datos
homicidios_hechos["Hora_entero"].apply(type).value_counts()

Hora_entero
<class 'int'>    695
<class 'str'>      1
Name: count, dtype: int64

Observamos que es un solo datos que presenta esta caracteristica

In [30]:
# Se imputa la hora moda al dato faltante
homicidios_hechos['Hora_entero'] = homicidios_hechos['Hora_entero'].apply(lambda x: int(hora_moda.hour) if x == "SD" else x)

In [31]:
#Verificamos
homicidios_hechos["Hora_entero"].apply(type).value_counts()

Hora_entero
<class 'int'>    696
Name: count, dtype: int64

Obtenemos en esta columna los 696 datos del mismo tipo

Verificamos el registro completado

In [32]:
# Se verifica el registro completado
homicidios_hechos[homicidios_hechos["Id_hecho"]=='2019-0103']

Id_hecho  Cantidad_victimas      Fecha   Año  Mes  Dia      Hora  \
518  2019-0103                  1 2019-12-18  2019   12   18  09:00:00   

     Hora_entero          Lugar_del_hecho Tipo_de_calle           Calle Cruce  \
518            9  PAZ, GRAL. AV. Y GRIVEO      GRAL PAZ  PAZ, GRAL. AV.    SI   

       Dirección_normalizada  Comuna                               Xy_(caba)  \
518  PAZ, GRAL. AV. y GRIVEO      11  Point (94643.11254058 103831.57115061)   

            Pos_x         Pos_y Participantes Victima Acusado  
518  -58.52169422  -34.59471640     MOTO-MOTO    MOTO    MOTO

Calle

esta columna presenta dos tipos de datos, string y float, al mismo timepo presenta un valor nulo, estos valores pueden tener relacion

In [33]:
# Verificamos
homicidios_hechos["Calle"].apply(type).value_counts()

Calle
<class 'str'>      695
<class 'float'>      1
Name: count, dtype: int64

Veremos cual es el dato de tipo float

In [34]:
homicidios_hechos[homicidios_hechos["Calle"].apply(lambda x: isinstance(x, float))]

Id_hecho  Cantidad_victimas      Fecha   Año  Mes  Dia      Hora  \
119  2016-0151                  1 2016-11-18  2016   11   18  20:35:00   

     Hora_entero Lugar_del_hecho Tipo_de_calle Calle Cruce  \
119           20              SD         CALLE   NaN    NO   

    Dirección_normalizada  Comuna    Xy_(caba) Pos_x Pos_y Participantes  \
119                    SD       0  Point (. .)     .     .     PEATON-SD   

    Victima Acusado  
119  PEATON      SD

Podemos observar que el dato de tipo float de la columna es un dato nulo o NaN, de esta manera nos damos cuenta de que es el mismo dato

In [35]:
# Se imputa SD al dato faltante
homicidios_hechos['Calle'].fillna('SD', inplace=True)

In [36]:
# verificamos
homicidios_hechos[homicidios_hechos["Id_hecho"]== "2016-0151"]

Id_hecho  Cantidad_victimas      Fecha   Año  Mes  Dia      Hora  \
119  2016-0151                  1 2016-11-18  2016   11   18  20:35:00   

     Hora_entero Lugar_del_hecho Tipo_de_calle Calle Cruce  \
119           20              SD         CALLE    SD    NO   

    Dirección_normalizada  Comuna    Xy_(caba) Pos_x Pos_y Participantes  \
119                    SD       0  Point (. .)     .     .     PEATON-SD   

    Victima Acusado  
119  PEATON      SD

### Victima y Acusado

Verificamos las columnas

In [37]:
homicidios_hechos["Victima"].unique()

array(['MOTO', 'AUTO', 'PEATON', 'SD', 'CARGAS', 'BICICLETA', 'PASAJEROS',
       'MOVIL', 'OBJETO FIJO', 'PEATON_MOTO'], dtype=object)

In [38]:
homicidios_hechos["Acusado"].unique()

array(['AUTO', 'PASAJEROS', 'SD', 'OBJETO FIJO', 'CARGAS', 'MOTO',
       'MULTIPLE', 'OTRO', 'BICICLETA', 'TREN'], dtype=object)

Observamos que en la columna victima segun el diccionario de datos nos deberia aparecer algunos datos como otro ya que asi lo manifiesta, pero en el dataframe nos aparece dos valores que no se ven reflejados en el diccionario como lo son Objeto fijo y peaton moto

In [39]:
# Verificamos datos
homicidios_hechos[homicidios_hechos['Victima'].isin(['OBJETO FIJO', 'PEATON_MOTO'])]

Id_hecho  Cantidad_victimas      Fecha   Año  Mes  Dia      Hora  \
230  2017-0108                  2 2017-09-02  2017    9    2  04:53:08   
583  2020-0063                  2 2020-12-05  2020   12    5  07:10:00   

     Hora_entero           Lugar_del_hecho Tipo_de_calle           Calle  \
230            4   AV. GRAL. PAZ Y MACHAIN      GRAL PAZ  PAZ, GRAL. AV.   
583            7  NUEVA YORK Y ALTA GRACIA         CALLE      NUEVA YORK   

    Cruce     Dirección_normalizada  Comuna  \
230    SI  PAZ, GRAL. AV. y MACHAIN      12   
583    SI  NUEVA YORK y ALTA GRACIA      11   

                                  Xy_(caba)         Pos_x         Pos_y  \
230  Point (97098.48468623 109019.96106626)  -58.49491054  -34.54795581   
583  Point (94080.62190808 102083.62453795)  -58.52783814  -34.61047001   

        Participantes      Victima Acusado  
230  AUTO-OBJETO FIJO  OBJETO FIJO    AUTO  
583  PEATON_MOTO-MOTO  PEATON_MOTO    MOTO

Solo son dos registros que presentan esta anomalia y seria uno de cada uno

In [40]:
homicidios_hechos['Victima'] = homicidios_hechos['Victima'].replace({'OBJETO FIJO':'OTRO', 'PEATON_MOTO':'OTRO'})

Cambiamos sus valores por el valor otro como lo indica el diccionario de datos

In [41]:
# verificamos valores cambiados
homicidios_hechos[homicidios_hechos["Id_hecho"]== "2017-0108"]

Id_hecho  Cantidad_victimas      Fecha   Año  Mes  Dia      Hora  \
230  2017-0108                  2 2017-09-02  2017    9    2  04:53:08   

     Hora_entero          Lugar_del_hecho Tipo_de_calle           Calle Cruce  \
230            4  AV. GRAL. PAZ Y MACHAIN      GRAL PAZ  PAZ, GRAL. AV.    SI   

        Dirección_normalizada  Comuna                               Xy_(caba)  \
230  PAZ, GRAL. AV. y MACHAIN      12  Point (97098.48468623 109019.96106626)   

            Pos_x         Pos_y     Participantes Victima Acusado  
230  -58.49491054  -34.54795581  AUTO-OBJETO FIJO    OTRO    AUTO

### Cooordenadas

Podemos observar durante el analisis que existen algunos datos que no aparecen como nulos den el dataframe, pero que estan como "." en las columnas de coordenadas, los cuales serian un gran problema a la hora de interacuar con essos datos, por lo tanto se decide reemplazarlos por el valor 0

In [42]:
# Verificamos datos
homicidios_hechos[homicidios_hechos['Xy_(caba)']=='Point (. .)']

Id_hecho  Cantidad_victimas      Fecha   Año  Mes  Dia      Hora  \
35   2016-0049                  1 2016-04-17  2016    4   17  00:00:00   
38   2016-0052                  1 2016-04-20  2016    4   20  20:00:00   
71   2016-0096                  1 2016-07-25  2016    7   25  07:00:00   
106  2016-0136                  1 2016-10-25  2016   10   25  00:00:00   
119  2016-0151                  1 2016-11-18  2016   11   18  20:35:00   
139  2016-0174                  1 2016-12-27  2016   12   27  00:00:00   
176  2017-0042                  1 2017-04-10  2017    4   10  09:00:00   
180  2017-0050                  2 2017-04-28  2017    4   28  11:08:08   
181  2017-0051                  1 2017-05-01  2017    5    1  03:47:47   
256  2017-0140                  1 2017-11-19  2017   11   19  23:22:17   
313  2018-0039                  1 2018-04-21  2018    4   21  22:15:00   
546  2020-0026                  1 2020-05-17  2020    5   17  06:40:00   
559  2020-0039                  1 2020-09-01  2020    9    1  19:17:42   
621  2021-0023                  1 2021-03-01  2021    3    1  09:20:00   

     Hora_entero                                    Lugar_del_hecho  \
35             0  AUTOPISTA 1 SUR PRESIDENTE ARTURO FRONDIZI KM....   
38            20                         AUTOPISTA LUGONES PK 10000   
71             7       AUTOPISTA DELLEPIANE LUIS TTE. GRAL. KM. 2,3   
106            0                   AU BUENOS AIRES - LA PLATA KM. 4   
119           20                                                 SD   
139            0                               AUTOPISTA 25 DE MAYO   
176            9                      AV. LEOPOLDO LUGONES PKM 6900   
180           11            AU PERITO MORENO Y RAMAL ENLACE AU1/AU6   
181            3                                 AU DELLEPIANE 2400   
256           23                        AU ARTURO FRONDIZI PKM 3100   
313           22                           AUTOPISTA LUGONES KM 4.7   
546            6                       LUGONES, LEOPOLDO AV. KM 6,1   
559           19                                    MURGUIONDO 2700   
621            9                   AU BUENOS AIRES LA PLATA  KM 4,5   

    Tipo_de_calle                                       Calle Cruce  \
35      AUTOPISTA  AUTOPISTA 1 SUR PRESIDENTE ARTURO FRONDIZI    NO   
38      AUTOPISTA                       LUGONES, LEOPOLDO AV.    NO   
71      AUTOPISTA        AUTOPISTA DELLEPIANE LUIS TTE. GRAL.    NO   
106     AUTOPISTA           AUTOPISTA BUENOS AIRES - LA PLATA    NO   
119         CALLE                                          SD    NO   
139     AUTOPISTA                        AUTOPISTA 25 DE MAYO    NO   
176      GRAL PAZ                       LUGONES, LEOPOLDO AV.    NO   
180     AUTOPISTA                     AUTOPISTA PERITO MORENO    NO   
181     AUTOPISTA        AUTOPISTA DELLEPIANE LUIS TTE. GRAL.    NO   
256     AUTOPISTA  AUTOPISTA 1 SUR PRESIDENTE ARTURO FRONDIZI    NO   
313     AUTOPISTA                       LUGONES, LEOPOLDO AV.    NO   
546     AUTOPISTA                       LUGONES, LEOPOLDO AV.    NO   
559         CALLE                                  MURGUIONDO    NO   
621     AUTOPISTA           AUTOPISTA BUENOS AIRES - LA PLATA    NO   

                          Dirección_normalizada  Comuna    Xy_(caba)  \
35   AUTOPISTA 1 SUR PRESIDENTE ARTURO FRONDIZI       4  Point (. .)   
38                                           SD      13  Point (. .)   
71         AUTOPISTA DELLEPIANE LUIS TTE. GRAL.       8  Point (. .)   
106                                          SD       4  Point (. .)   
119                                          SD       0  Point (. .)   
139                        AUTOPISTA 25 DE MAYO       0  Point (. .)   
176                       LUGONES, LEOPOLDO AV.      14  Point (. .)   
180                                          SD       9  Point (. .)   
181                                          SD       7  Point (. .)   
256  AUTOPISTA 1 SUR PRESIDENTE ARTURO FRON

Reemplazamos datos

In [43]:
homicidios_hechos['Pos_x'] = homicidios_hechos['Pos_x'].replace('.', 0)
homicidios_hechos['Pos_y'] = homicidios_hechos['Pos_y'].replace('.', 0)
homicidios_hechos['Xy_(caba)'] = homicidios_hechos['Xy_(caba)'].replace('Point (. .)', 0)

## Pestaña Victimas

In [44]:
# se exytraen los datos
homicidios_victimas = pd.read_excel("../Datos/homicidios.xlsx", sheet_name="VICTIMAS")
homicidios_victimas

ID_hecho      FECHA  AAAA  MM  DD                   ROL VICTIMA  \
0    2016-0001 2016-01-01  2016   1   1             CONDUCTOR    MOTO   
1    2016-0002 2016-01-02  2016   1   2             CONDUCTOR    AUTO   
2    2016-0003 2016-01-03  2016   1   3             CONDUCTOR    MOTO   
3    2016-0004 2016-01-10  2016   1  10             CONDUCTOR    MOTO   
4    2016-0005 2016-01-21  2016   1  21             CONDUCTOR    MOTO   
..         ...        ...   ...  ..  ..                   ...     ...   
712  2021-0092 2021-12-12  2021  12  12                PEATON  PEATON   
713  2021-0093 2021-12-13  2021  12  13  PASAJERO_ACOMPAÑANTE    MOTO   
714  2021-0094 2021-12-20  2021  12  20  PASAJERO_ACOMPAÑANTE    MOTO   
715  2021-0095 2021-12-30  2021  12  30             CONDUCTOR    MOTO   
716  2021-0096 2021-12-15  2021  12  15             CONDUCTOR    AUTO   

          SEXO EDAD  FECHA_FALLECIMIENTO  
0    MASCULINO   19  2016-01-01 00:00:00  
1    MASCULINO   70  2016-01-02 00:00:00  
2    MASCULINO   30  2016-01-03 00:00:00  
3    MASCULINO   18                   SD  
4    MASCULINO   29  2016-02-01 00:00:00  
..         ...  ...                  ...  
712   FEMENINO   50  2021-12-12 00:00:00  
713   FEMENINO   18  2021-12-18 00:00:00  
714   FEMENINO   43  2021-12-20 00:00:00  
715  MASCULINO   27  2022-01-02 00:00:00  
716  MASCULINO   60  2021-12-20 00:00:00  

[717 rows x 10 columns]

Este conjunto de datos contiene 10 cplumnas y 717 registros en total

In [45]:
# Columnas del dataset
homicidios_victimas.columns

Index(['ID_hecho', 'FECHA', 'AAAA', 'MM', 'DD', 'ROL', 'VICTIMA', 'SEXO',
       'EDAD', 'FECHA_FALLECIMIENTO'],
      dtype='object')

Podemos Observar que algunas columnas tienen nobres en mayusculas y otras en minusculas, asi que normalizaremos como en el dataframe anterior

In [46]:
# Se coloca la primera en mayúscula
homicidios_victimas.columns = [x.capitalize() for x in homicidios_victimas.columns]

In [47]:
# Se renombran algunas columnas
homicidios_victimas = homicidios_victimas.rename(columns={'Aaaa':'Año',
                                                      'Mm':'Mes',
                                                      'Dd':'Dia'})
homicidios_victimas.columns

Index(['Id_hecho', 'Fecha', 'Año', 'Mes', 'Dia', 'Rol', 'Victima', 'Sexo',
       'Edad', 'Fecha_fallecimiento'],
      dtype='object')

### Datos nulos

In [48]:
funciones_etl.verificar_datos(homicidios_victimas)

columna                                          tipo_dato  \
0             Id_hecho                                    [<class 'str'>]   
1                Fecha  [<class 'pandas._libs.tslibs.timestamps.Timest...   
2                  Año                                    [<class 'int'>]   
3                  Mes                                    [<class 'int'>]   
4                  Dia                                    [<class 'int'>]   
5                  Rol                                    [<class 'str'>]   
6              Victima                                    [<class 'str'>]   
7                 Sexo                                    [<class 'str'>]   
8                 Edad                     [<class 'int'>, <class 'str'>]   
9  Fecha_fallecimiento       [<class 'datetime.datetime'>, <class 'str'>]   

   datos_nulos  porcentaje_nulos  porcentaje_no_nulos  
0            0               0.0                100.0  
1            0               0.0                100.0  
2            0               0.0                100.0  
3            0               0.0                100.0  
4            0               0.0                100.0  
5            0               0.0                100.0  
6            0               0.0                100.0  
7            0               0.0                100.0  
8            0               0.0                100.0  
9            0               0.0                100.0

No presenta daos nulos, sin embargo dos columnas tienen dos tipos de datos, la columnas Edad y la columna Fecha_fallecimiento

### Dupliacdos

Valores duplicados a nivel general

In [49]:
homicidios_victimas[homicidios_victimas.duplicated()]

Empty DataFrame
Columns: [Id_hecho, Fecha, Año, Mes, Dia, Rol, Victima, Sexo, Edad, Fecha_fallecimiento]
Index: []

No posee valores dupliacos a nivel general

Ducplicados por Id

In [50]:
funciones_etl.valores_duplicados(homicidios_victimas, "Id_hecho")

41

Tenemos 41 datos duplicados por id, procedemos a observar esos datos

In [51]:
funciones_etl.ver_duplicados(homicidios_victimas, "Id_hecho")

Id_hecho      Fecha   Año  Mes  Dia                   Rol Victima  \
29   2016-0041 2016-03-29  2016    3   29             CONDUCTOR    MOTO   
30   2016-0041 2016-03-29  2016    3   29  PASAJERO_ACOMPAÑANTE    MOTO   
98   2016-0126 2016-09-18  2016    9   18             CONDUCTOR    AUTO   
99   2016-0126 2016-09-18  2016    9   18  PASAJERO_ACOMPAÑANTE    AUTO   
163  2017-0026 2017-02-26  2017    2   26  PASAJERO_ACOMPAÑANTE    AUTO   
164  2017-0026 2017-02-26  2017    2   26             CONDUCTOR    AUTO   
173  2017-0035 2017-03-23  2017    3   23             CONDUCTOR    AUTO   
174  2017-0035 2017-03-23  2017    3   23  PASAJERO_ACOMPAÑANTE    AUTO   
175  2017-0035 2017-03-23  2017    3   23  PASAJERO_ACOMPAÑANTE    AUTO   
177  2017-0036 2017-03-29  2017    3   29             CONDUCTOR    MOTO   
176  2017-0036 2017-03-29  2017    3   29  PASAJERO_ACOMPAÑANTE    MOTO   
186  2017-0050 2017-04-28  2017    4   28             CONDUCTOR    MOTO   
187  2017-0050 2017-04-28  2017    4   28  PASAJERO_ACOMPAÑANTE    MOTO   
237  2017-0108 2017-09-02  2017    9    2  PASAJERO_ACOMPAÑANTE    AUTO   
238  2017-0108 2017-09-02  2017    9    2  PASAJERO_ACOMPAÑANTE    AUTO   
242  2017-0112 2017-09-10  2017    9   10  PASAJERO_ACOMPAÑANTE    AUTO   
243  2017-0112 2017-09-10  2017    9   10  PASAJERO_ACOMPAÑANTE    AUTO   
253  2017-0126 2017-10-14  2017   10   14  PASAJERO_ACOMPAÑANTE    AUTO   
254  2017-0126 2017-10-14  2017   10   14             CONDUCTOR    AUTO   
278  2017-0153 2017-12-11  2017   12   11             CONDUCTOR    MOTO   
277  2017-0153 2017-12-11  2017   12   11  PASAJERO_ACOMPAÑANTE    MOTO   
300  2018-0015 2018-02-06  2018    2    6             CONDUCTOR    MOTO   
301  2018-0015 2018-02-06  2018    2    6  PASAJERO_ACOMPAÑANTE    MOTO   
312  2018-0026 2018-03-09  2018    3    9             CONDUCTOR   MOVIL   
313  2018-0026 2018-03-09  2018    3    9  PASAJERO_ACOMPAÑANTE   MOVIL   
325  2018-0038 2018-04-20  2018    4   20                PEATON  PEATON   
326  2018-0038 2018-04-20  2018    4   20                PEATON  PEATON   
373  2018-0085 2018-08-03  2018    8    3  PASAJERO_ACOMPAÑANTE    MOTO   
374  2018-0085 2018-08-03  2018    8    3             CONDUCTOR    MOTO   
388  2018-0098 2018-09-09  2018    9    9  PASAJERO_ACOMPAÑANTE    AUTO   
387  2018-0098 2018-09-09  2018    9    9             CONDUCTOR    AUTO   
418  2018-0128 2018-11-25  2018   11   25  PASAJERO_ACOMPAÑANTE    AUTO   
419  2018-0128 2018-11-25  2018   11   25  PASAJERO_ACOMPAÑANTE    AUTO   
444  2019-0010 2019-02-02  2019    2    2                PEATON  PEATON   
445  2019-0010 2019-02-02  2019    2    2                PEATON  PEATON   
576  2020-0038 2020-08-30  2020    8   30             CONDUCTOR    AUTO   
577  2020-0038 2020-08-30  2020    8   30             CONDUCTOR    AUTO   
602  2020-0063 2020-12-05  2020   12    5                PEATON  PEATON   
603  2020-0063 2020-12-05  2020   12    5  PASAJERO_ACOMPAÑANTE    MOTO   
616  2020-0077 2020-12-25  2020   12   25             CONDUCTOR    MOTO   
617  2020-0077 2020-12-25  2020   12   25  PASAJERO_ACOMPAÑANTE    MOTO   

          Sexo Edad  Fecha_fallecimiento  
29   MASCULINO   54  2016-03-29 00:00:00  
30   MASCULINO   SD  2016-03-30 00:00:00  
98   MASCULINO   37                   SD  
99   MASCULINO   60                   SD  
163   FEMENINO   23  2017-02-26 00:00:00  
164  MASCULINO   19  2017-02-26 00:00:00  
173  MASCULINO   28  2017-03-23 00:00:00  
174  MASCULINO   32  2017-03-23 00:00:00  
175  MASCULINO   30  2017-03-23 00:00:00  
177  MASCULINO   20  2017-03-29 00:00:00  
176   FEMENINO   SD  2017-03-29 00:00:00  
186  MASCULINO   46  2017-04-28 00:00:00  
187  MASCULINO   16  2017-04-28 00:00:00  
237  MASCULINO   17  2017-09-02 00:00:00  
238  MASCULINO   10  2017-09-02 00:00:00  
242  MASCULINO    1                   SD  
243  MASCULINO   SD  2017-09-13 00:00:00  
253  MASCULINO   39                   SD  
254  MASCULINO   36  2017-10-14 00:00:

A pesar de que estos datos comparten el mismo id obsercamos que no se trata de la misma victima, solo se trata que es del mismo hecho, por lo tanto estan repetidos como tal

### Edad

Oservamos las cantidades por tipos de datos

In [52]:
homicidios_victimas["Edad"].apply(type).value_counts()

Edad
<class 'int'>    664
<class 'str'>     53
Name: count, dtype: int64

664 de sus datos son de tipo int, lo cual es coherente, pero 53 datos son de tipo string, los cuals vamos a observar

In [53]:
homicidios_victimas[homicidios_victimas["Edad"].apply(lambda x: isinstance(x, str))]

Id_hecho      Fecha   Año  Mes  Dia                   Rol    Victima  \
30   2016-0041 2016-03-29  2016    3   29  PASAJERO_ACOMPAÑANTE       MOTO   
33   2016-0045 2016-04-11  2016    4   11             CONDUCTOR       MOTO   
35   2016-0048 2016-04-15  2016    4   15                PEATON     PEATON   
36   2016-0049 2016-04-17  2016    4   17                    SD         SD   
39   2016-0052 2016-04-20  2016    4   20                    SD       MOTO   
55   2016-0077 2016-06-13  2016    6   13                PEATON     PEATON   
63   2016-0085 2016-06-29  2016    6   29                    SD       MOTO   
72   2016-0096 2016-07-25  2016    7   25             CONDUCTOR       MOTO   
89   2016-0115 2016-09-02  2016    9    2                    SD         SD   
93   2016-0119 2016-09-04  2016    9    4  PASAJERO_ACOMPAÑANTE         SD   
108  2016-0136 2016-10-25  2016   10   25             CONDUCTOR       MOTO   
119  2016-0148 2016-11-16  2016   11   16             CONDUCTOR       MOTO   
121  2016-0151 2016-11-18  2016   11   18                PEATON     PEATON   
138  2016-0171 2016-12-25  2016   12   25             CONDUCTOR       MOTO   
141  2016-0174 2016-12-27  2016   12   27                    SD         SD   
142  2016-0175 2016-12-27  2016   12   27                PEATON     PEATON   
165  2017-0027 2017-02-26  2017    2   26                PEATON     PEATON   
176  2017-0036 2017-03-29  2017    3   29  PASAJERO_ACOMPAÑANTE       MOTO   
182  2017-0042 2017-04-10  2017    4   10             CONDUCTOR       MOTO   
225  2017-0093 2017-07-25  2017    7   25                PEATON     PEATON   
243  2017-0112 2017-09-10  2017    9   10  PASAJERO_ACOMPAÑANTE       AUTO   
275  2017-0151 2017-12-04  2017   12    4                PEATON     PEATON   
277  2017-0153 2017-12-11  2017   12   11  PASAJERO_ACOMPAÑANTE       MOTO   
299  2018-0014 2018-01-31  2018    1   31                PEATON     PEATON   
328  2018-0040 2018-04-27  2018    4   27                PEATON     PEATON   
359  2018-0071 2018-07-10  2018    7   10             CONDUCTOR       MOTO   
367  2018-0079 2018-07-23  2018    7   23                PEATON     PEATON   
371  2018-0083 2018-08-02  2018    8    2             CONDUCTOR       MOTO   
377  2018-0088 2018-08-17  2018    8   17             CONDUCTOR       MOTO   
391  2018-0101 2018-09-21  2018    9   21             CONDUCTOR       AUTO   
421  2018-0130 2018-12-03  2018   12    3             CONDUCTOR       MOTO   
444  2019-0010 2019-02-02  2019    2    2                PEATON     PEATON   
446  2019-0011 2019-02-10  2019    2   10                PEATON     PEATON   
454  2019-0019 2019-03-06  2019    3    6             CONDUCTOR       MOTO   
457  2019-0022 2019-03-15  2019    3   15  PASAJERO_ACOMPAÑANTE       AUTO   
461  2019-0026 2019-03-28  2019    3   28                PEATON     PEATON   
463  2019-0028 2019-04-06  2019    4    6                PEATON     PEATON   
469  2019-0034 2019-04-21  2019    4   21             CONDUCTOR       MOTO   
479  2019-0043 2019-05-21  2019    5   21              CICLISTA  BICICLETA   
498  2019-0063 2019-08-04  2019    8    4  PASAJERO_ACOMPAÑANTE       MOTO   
506  2019-0071 2019-08-27  2019    8   27                PEATON     PEATON   
529  2019-0094 2019-11-18  2019   11   18  PASAJERO_ACOMPAÑANTE       MOTO   
530  2019-0095 2019-11-20  2019   11   20             CONDUCTOR       MOTO   
548  2020-0010 2020-02-08  2020    2    8             CONDUCTOR       MOTO   
552  2020-0014 2020-02-27  2020    2   27             CONDUCTOR       MOTO   
555  2020-0017 2020-03-05  2020    3    5                PEATON     PEATON   
563  2020-0025 2020-05-12  2020    5   12                PEATON     PEATON   
572  2020-0034 2020-08-02  2020    8    2                PEATON     PEATON   
579  2020-0040 2020-09-12  2020    9   12             CONDUCTOR       AUTO   
631  2021-0012 2021-01-29  2021    1   29                PEATON     PEATON   
639  2021-0020 2021-02-21  2

Observamos 53 datos nulos en la columna edad, pero tambien observamos datos nulos en fecha_fallecimiento, rol, sexo

Para poder imputar los valores de esdad utilizaremos medidas de tendencia central, pero para ello debemos clasificar segun el sexo, lo cual nos lleva a imputar primero la columna sexo y luego la columna edad

In [54]:
# Se imputa segun el sexo
funciones_etl.imputa_valor_frecuente(homicidios_victimas, "Sexo")

El valor mas frecuente es: MASCULINO


De esta manera con los valores de la columna sexo ya impoutados podemos clasificar y poder imputar segun la media de edad

In [55]:
# Se imputa segun la media de edad por sexo
funciones_etl.imputa_edad_media_segun_sexo(homicidios_victimas)

La edad promedio de Femenino es 51 y de Masculino es 40


### Rol

procedemos a verificar los valores en la columna rol que sean "SD"

In [56]:
nulos_rol = homicidios_victimas[homicidios_victimas["Rol"]== "SD"]
nulos_rol

Id_hecho      Fecha   Año  Mes  Dia Rol Victima       Sexo  Edad  \
36   2016-0049 2016-04-17  2016    4   17  SD      SD  MASCULINO    39   
39   2016-0052 2016-04-20  2016    4   20  SD    MOTO  MASCULINO    39   
63   2016-0085 2016-06-29  2016    6   29  SD    MOTO  MASCULINO    39   
77   2016-0101 2016-08-07  2016    8    7  SD      SD  MASCULINO    67   
89   2016-0115 2016-09-02  2016    9    2  SD      SD  MASCULINO    39   
141  2016-0174 2016-12-27  2016   12   27  SD      SD  MASCULINO    39   
167  2017-0029 2017-03-07  2017    3    7  SD      SD  MASCULINO    34   
208  2017-0074 2017-06-04  2017    6    4  SD      SD  MASCULINO    70   
221  2017-0089 2017-07-13  2017    7   13  SD      SD  MASCULINO    23   
280  2017-0155 2017-12-12  2017   12   12  SD      SD  MASCULINO    77   
390  2018-0100 2018-09-20  2018    9   20  SD    MOTO   FEMENINO    33   

     Fecha_fallecimiento  
36                    SD  
39                    SD  
63                    SD  
77                    SD  
89                    SD  
141                   SD  
167  2017-03-07 00:00:00  
208  2017-06-04 00:00:00  
221                   SD  
280                   SD  
390  2018-09-20 00:00:00

In [57]:
len(nulos_rol)

11

Encontramos 11 datos nulos en esta columna

In [58]:
# Imputamos por valor frecuente
funciones_etl.imputa_valor_frecuente(homicidios_victimas, "Rol")

El valor mas frecuente es: CONDUCTOR


### Victima

Observamos valores nulos

In [59]:
nulos_victima = homicidios_victimas[homicidios_victimas["Victima"]== "SD"]
nulos_victima

Id_hecho      Fecha   Año  Mes  Dia                   Rol Victima  \
36   2016-0049 2016-04-17  2016    4   17             CONDUCTOR      SD   
77   2016-0101 2016-08-07  2016    8    7             CONDUCTOR      SD   
89   2016-0115 2016-09-02  2016    9    2             CONDUCTOR      SD   
93   2016-0119 2016-09-04  2016    9    4  PASAJERO_ACOMPAÑANTE      SD   
141  2016-0174 2016-12-27  2016   12   27             CONDUCTOR      SD   
167  2017-0029 2017-03-07  2017    3    7             CONDUCTOR      SD   
208  2017-0074 2017-06-04  2017    6    4             CONDUCTOR      SD   
221  2017-0089 2017-07-13  2017    7   13             CONDUCTOR      SD   
280  2017-0155 2017-12-12  2017   12   12             CONDUCTOR      SD   

          Sexo  Edad  Fecha_fallecimiento  
36   MASCULINO    39                   SD  
77   MASCULINO    67                   SD  
89   MASCULINO    39                   SD  
93    FEMENINO    50                   SD  
141  MASCULINO    39                   SD  
167  MASCULINO    34  2017-03-07 00:00:00  
208  MASCULINO    70  2017-06-04 00:00:00  
221  MASCULINO    23                   SD  
280  MASCULINO    77                   SD

In [60]:
len(nulos_victima)

9

Encontramos 9 valores nulos en la columna victima, los cuales imputaremos

In [61]:
funciones_etl.imputa_valor_frecuente(homicidios_victimas, "Victima")

El valor mas frecuente es: MOTO


### Fecha fallecimiento

Como el objetivo de este proyecto se centra en patrones y KPIS que sirvan para disminuir los accidentes fatales, se decide prescindir de la columna Fecha_fallecimiento, debido a que no aporta informacion reelevante en este contexto

In [62]:
homicidios_victimas = homicidios_victimas.drop("Fecha_fallecimiento", axis=1)
homicidios_victimas

Id_hecho      Fecha   Año  Mes  Dia                   Rol Victima  \
0    2016-0001 2016-01-01  2016    1    1             CONDUCTOR    MOTO   
1    2016-0002 2016-01-02  2016    1    2             CONDUCTOR    AUTO   
2    2016-0003 2016-01-03  2016    1    3             CONDUCTOR    MOTO   
3    2016-0004 2016-01-10  2016    1   10             CONDUCTOR    MOTO   
4    2016-0005 2016-01-21  2016    1   21             CONDUCTOR    MOTO   
..         ...        ...   ...  ...  ...                   ...     ...   
712  2021-0092 2021-12-12  2021   12   12                PEATON  PEATON   
713  2021-0093 2021-12-13  2021   12   13  PASAJERO_ACOMPAÑANTE    MOTO   
714  2021-0094 2021-12-20  2021   12   20  PASAJERO_ACOMPAÑANTE    MOTO   
715  2021-0095 2021-12-30  2021   12   30             CONDUCTOR    MOTO   
716  2021-0096 2021-12-15  2021   12   15             CONDUCTOR    AUTO   

          Sexo  Edad  
0    MASCULINO    19  
1    MASCULINO    70  
2    MASCULINO    30  
3    MASCULINO    18  
4    MASCULINO    29  
..         ...   ...  
712   FEMENINO    50  
713   FEMENINO    18  
714   FEMENINO    43  
715  MASCULINO    27  
716  MASCULINO    60  

[717 rows x 9 columns]

Ademas podemos observar que las columnas  'Fecha', 'Año', 'Mes', 'Día' y 'Víctima' contienen la misma informacion en ambos dataframes, por lo tanto tambien las eliminamos 

In [63]:
homicidios_victimas = homicidios_victimas.drop(["Fecha", "Año", "Mes", "Dia", "Victima"], axis=1)

Observamos el dataframe 

In [64]:
homicidios_victimas

Id_hecho                   Rol       Sexo  Edad
0    2016-0001             CONDUCTOR  MASCULINO    19
1    2016-0002             CONDUCTOR  MASCULINO    70
2    2016-0003             CONDUCTOR  MASCULINO    30
3    2016-0004             CONDUCTOR  MASCULINO    18
4    2016-0005             CONDUCTOR  MASCULINO    29
..         ...                   ...        ...   ...
712  2021-0092                PEATON   FEMENINO    50
713  2021-0093  PASAJERO_ACOMPAÑANTE   FEMENINO    18
714  2021-0094  PASAJERO_ACOMPAÑANTE   FEMENINO    43
715  2021-0095             CONDUCTOR  MASCULINO    27
716  2021-0096             CONDUCTOR  MASCULINO    60

[717 rows x 4 columns]

Unimos los dataframes

In [65]:
homicidios = homicidios_victimas.merge(homicidios_hechos, on="Id_hecho", how="left")
homicidios

Id_hecho                   Rol       Sexo  Edad  Cantidad_victimas  \
0    2016-0001             CONDUCTOR  MASCULINO    19                  1   
1    2016-0002             CONDUCTOR  MASCULINO    70                  1   
2    2016-0003             CONDUCTOR  MASCULINO    30                  1   
3    2016-0004             CONDUCTOR  MASCULINO    18                  1   
4    2016-0005             CONDUCTOR  MASCULINO    29                  1   
..         ...                   ...        ...   ...                ...   
712  2021-0092                PEATON   FEMENINO    50                  1   
713  2021-0093  PASAJERO_ACOMPAÑANTE   FEMENINO    18                  1   
714  2021-0094  PASAJERO_ACOMPAÑANTE   FEMENINO    43                  1   
715  2021-0095             CONDUCTOR  MASCULINO    27                  1   
716  2021-0096             CONDUCTOR  MASCULINO    60                  1   

         Fecha   Año  Mes  Dia      Hora  ...                         Calle  \
0   2016-01-01  2016    1    1  04:00:00  ...              PIEDRA BUENA AV.   
1   2016-01-02  2016    1    2  01:15:00  ...                PAZ, GRAL. AV.   
2   2016-01-03  2016    1    3  07:00:00  ...                ENTRE RIOS AV.   
3   2016-01-10  2016    1   10  00:00:00  ...                LARRAZABAL AV.   
4   2016-01-21  2016    1   21  05:20:00  ...                  SAN JUAN AV.   
..         ...   ...  ...  ...       ...  ...                           ...   
712 2021-12-12  2021   12   12  06:20:00  ...                 RIVADAVIA AV.   
713 2021-12-13  2021   12   13  17:10:00  ...                   RIESTRA AV.   
714 2021-12-20  2021   12   20  01:10:00  ...  DELLEPIANE, LUIS, TTE. GRAL.   
715 2021-12-30  2021   12   30  00:43:00  ...                     GAONA AV.   
716 2021-12-15  2021   12   15  10:30:00  ...                PERON, EVA AV.   

    Cruce                              Dirección_normalizada Comuna  \
0      SI  PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...      8   
1      SI               PAZ, GRAL. AV. y DE LOS CORRALES AV.      9   
2      NO                                ENTRE RIOS AV. 2034      1   
3      SI          LARRAZABAL AV. y VILLEGAS, CONRADO, GRAL.      8   
4      SI            SAN JUAN AV. y SAENZ PEÑA, LUIS, PRES.      1   
..    ...                                                ...    ...   
712    SI                     RIVADAVIA AV. y PUEYRREDON AV.      3   
713    SI                                  RIESTRA AV. y MOM      7   
714    SI         DELLEPIANE, LUIS, TTE. GRAL. y LACARRA AV.      9   
715    SI                                GAONA AV. y TERRADA     11   
716    NO                                PERON, EVA AV. 4071      9   

                                   Xy_(caba)         Pos_x         Pos_y  \
0      Point (98896.78238426 93532.43437792)  -58.47533969  -34.68757022   
1      Point (95832.05571093 95505.41641999)  -58.50877521  -34.66977709   
2     Point (106684.29090040 99706.57687843)  -58.39040293  -34.63189362   
3      Point (99840.65224780 94269.16534422)  -58.46503904  -34.68092974   
4    Point (106980.32827929 100752.16915795)  -58.38718297  -34.62246630   
..                                       ...           ...           ...   
712  Point (105258.35368554 102122.93231400)  -58.40596860  -34.61011987   
713   Point (102728.60090138 98186.24929177)  -58.43353773  -34.64561636   
714    Point (99624.29795829 97569.69801131)  -58.46739825  -34.65117757   
715   Point (99116.45492358 101045.23284826)  -58.47293407  -34.61984745   
716    Point (99324.54463985 97676.26932409)  -58.47066794  -34.65021673   

      Participantes Victima    Acusado  
0         MOTO-AUTO    MOTO       AUTO  
1    AUTO-PASAJEROS    AUTO  PASAJEROS  
2         MOTO-AUTO    MOTO       AUTO  
3           MOTO-SD    MOTO         SD  
4    MOTO-PASAJEROS    MOTO  PASAJEROS  
..              ...     ...        ...  
712     PEATON-AUTO  PEATON       AUTO  
713       MOTO-AUTO    MOTO       AUTO  
714      

Verificamos nulos y tipos de datos

In [66]:
funciones_etl.verificar_datos(homicidios)

columna                                          tipo_dato  \
0                Id_hecho                                    [<class 'str'>]   
1                     Rol                                    [<class 'str'>]   
2                    Sexo                                    [<class 'str'>]   
3                    Edad                                    [<class 'int'>]   
4       Cantidad_victimas                                    [<class 'int'>]   
5                   Fecha  [<class 'pandas._libs.tslibs.timestamps.Timest...   
6                     Año                                    [<class 'int'>]   
7                     Mes                                    [<class 'int'>]   
8                     Dia                                    [<class 'int'>]   
9                    Hora                          [<class 'datetime.time'>]   
10            Hora_entero                                    [<class 'int'>]   
11        Lugar_del_hecho                                    [<class 'str'>]   
12          Tipo_de_calle                                    [<class 'str'>]   
13                  Calle                                    [<class 'str'>]   
14                  Cruce                                    [<class 'str'>]   
15  Dirección_normalizada                                    [<class 'str'>]   
16                 Comuna                                    [<class 'int'>]   
17              Xy_(caba)                     [<class 'str'>, <class 'int'>]   
18                  Pos_x                     [<class 'str'>, <class 'int'>]   
19                  Pos_y                     [<class 'str'>, <class 'int'>]   
20          Participantes                                    [<class 'str'>]   
21                Victima                                    [<class 'str'>]   
22                Acusado                                    [<class 'str'>]   

    datos_nulos  porcentaje_nulos  porcentaje_no_nulos  
0             0               0.0                100.0  
1             0               0.0                100.0  
2             0               0.0                100.0  
3             0               0.0                100.0  
4             0               0.0                100.0  
5             0               0.0                100.0  
6             0               0.0                100.0  
7             0               0.0                100.0  
8             0               0.0                100.0  
9             0               0.0                100.0  
10            0               0.0                100.0  
11            0               0.0                100.0  
12            0               0.0                100.0  
13            0               0.0                100.0  
14            0               0.0                100.0  
15            0               0.0                100.0  
16            0               0.0                100.0  
17            0               0.0                100.0  
18            0               0.0                100.0  
19            0               0.0                100.0  
20            0               0.0                100.0  
21            0               0.0                100.0  
22            0               0.0                100.0

### Semestre

Creamos una columna necesaria llamada semestre en donde clasificamos los siniestros, segun el mes donde fue realizado 

El etl esta completo procedemos a guardar el conjunto de datos en nuestra base de datos 

In [67]:
homicidios["Semestre"] = homicidios["Mes"].apply(funciones_etl.clasificar_semestre)

In [68]:
homicidios

Id_hecho                   Rol       Sexo  Edad  Cantidad_victimas  \
0    2016-0001             CONDUCTOR  MASCULINO    19                  1   
1    2016-0002             CONDUCTOR  MASCULINO    70                  1   
2    2016-0003             CONDUCTOR  MASCULINO    30                  1   
3    2016-0004             CONDUCTOR  MASCULINO    18                  1   
4    2016-0005             CONDUCTOR  MASCULINO    29                  1   
..         ...                   ...        ...   ...                ...   
712  2021-0092                PEATON   FEMENINO    50                  1   
713  2021-0093  PASAJERO_ACOMPAÑANTE   FEMENINO    18                  1   
714  2021-0094  PASAJERO_ACOMPAÑANTE   FEMENINO    43                  1   
715  2021-0095             CONDUCTOR  MASCULINO    27                  1   
716  2021-0096             CONDUCTOR  MASCULINO    60                  1   

         Fecha   Año  Mes  Dia      Hora  ...  Cruce  \
0   2016-01-01  2016    1    1  04:00:00  ...     SI   
1   2016-01-02  2016    1    2  01:15:00  ...     SI   
2   2016-01-03  2016    1    3  07:00:00  ...     NO   
3   2016-01-10  2016    1   10  00:00:00  ...     SI   
4   2016-01-21  2016    1   21  05:20:00  ...     SI   
..         ...   ...  ...  ...       ...  ...    ...   
712 2021-12-12  2021   12   12  06:20:00  ...     SI   
713 2021-12-13  2021   12   13  17:10:00  ...     SI   
714 2021-12-20  2021   12   20  01:10:00  ...     SI   
715 2021-12-30  2021   12   30  00:43:00  ...     SI   
716 2021-12-15  2021   12   15  10:30:00  ...     NO   

                                 Dirección_normalizada Comuna  \
0    PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...      8   
1                 PAZ, GRAL. AV. y DE LOS CORRALES AV.      9   
2                                  ENTRE RIOS AV. 2034      1   
3            LARRAZABAL AV. y VILLEGAS, CONRADO, GRAL.      8   
4              SAN JUAN AV. y SAENZ PEÑA, LUIS, PRES.      1   
..                                                 ...    ...   
712                     RIVADAVIA AV. y PUEYRREDON AV.      3   
713                                  RIESTRA AV. y MOM      7   
714         DELLEPIANE, LUIS, TTE. GRAL. y LACARRA AV.      9   
715                                GAONA AV. y TERRADA     11   
716                                PERON, EVA AV. 4071      9   

                                   Xy_(caba)         Pos_x         Pos_y  \
0      Point (98896.78238426 93532.43437792)  -58.47533969  -34.68757022   
1      Point (95832.05571093 95505.41641999)  -58.50877521  -34.66977709   
2     Point (106684.29090040 99706.57687843)  -58.39040293  -34.63189362   
3      Point (99840.65224780 94269.16534422)  -58.46503904  -34.68092974   
4    Point (106980.32827929 100752.16915795)  -58.38718297  -34.62246630   
..                                       ...           ...           ...   
712  Point (105258.35368554 102122.93231400)  -58.40596860  -34.61011987   
713   Point (102728.60090138 98186.24929177)  -58.43353773  -34.64561636   
714    Point (99624.29795829 97569.69801131)  -58.46739825  -34.65117757   
715   Point (99116.45492358 101045.23284826)  -58.47293407  -34.61984745   
716    Point (99324.54463985 97676.26932409)  -58.47066794  -34.65021673   

      Participantes Victima    Acusado Semestre  
0         MOTO-AUTO    MOTO       AUTO        1  
1    AUTO-PASAJEROS    AUTO  PASAJEROS        1  
2         MOTO-AUTO    MOTO       AUTO        1  
3           MOTO-SD    MOTO         SD        1  
4    MOTO-PASAJEROS    MOTO  PASAJEROS        1  
..              ...     ...        ...      ...  
712     PEATON-AUTO  PEATON       AUTO        2  
713       MOTO-AUTO    MOTO       AUTO        2  
714       MOTO-AUTO    MOTO       AUTO        2  
715     MOTO-CARGAS    MOTO     CARGAS        2  
716     AUTO-CARGAS    AUTO     CARGAS        2  

[717 rows x 24 columns]

In [69]:
homicidios.to_sql("homicidios_hecho", con=conexion_bd.engine, index= False, if_exists="replace")

717